### ISSUE TITLE
Refactor pipelines

### ISSUE THREAD

**Issue by frazane** [MeteoSwiss/mlpp-features#15]
I propose to do a major refactoring of all pipelines based on the following principles:

### Avoid computing the same values more than once
There are two complementary ways to accomplish this:

First, we can reuse pipelines that are already in the output dataset with the `@reuse` decorator that is currently implemented
It's for cases like this: 

if `t_d` and `t` have already been computed, they will be reused.
```python
@reuse
@asarray
def water_vapor_pressure_ensavg(
    data: Dict[str, xr.Dataset], stations, reftimes, leadtimes, **kwargs
) -> xr.DataArray:
    """
    Ensemble mean of water vapor partial pressure
    """
    # try/except block necessary to expose all the required input data
    try:
        data["nwp"]["dew_point_temperature"]
        data["nwp"]["air_temperature"]
    except KeyError:
        raise KeyError(["dew_point_temperature", "air_temperature"])

    t_d = dew_point_temperature_ensavg(data, stations, reftimes, leadtimes, **kwargs)
    t = air_temperature_ensavg(data, stations, reftimes, leadtimes, **kwargs)
    return calc.water_vapor_pressure_from_t_and_td(t, t_d)
```

Second, we can divide pipelines in smaller steps, and cache those steps that are needed by multiple pipelines.

```python
@callonce
def eastward_wind(
    data: Dict[str, xr.Dataset], stations, reftimes, leadtimes, **kwargs
):
    return (
        data["nwp"]
        .preproc.get("eastward_wind")
        .preproc.interp(stations)
        .preproc.align_time(reftimes, leadtimes)
        .astype("float32")
        )   

@callonce
def northward_wind(
    data: Dict[str, xr.Dataset], stations, reftimes, leadtimes, **kwargs
):
    return (
        data["nwp"]
        .preproc.get("northward_wind")
        .preproc.interp(stations)
        .preproc.align_time(reftimes, leadtimes)
        .astype("float32")
        )   

@callonce
def wind_speed(
    data: Dict[str, xr.Dataset], stations, reftimes, leadtimes, **kwargs
):
    u = eastward_wind(data, stations, reftimes, leadtimes, **kwargs)
    v = northward_wind(data, stations, reftimes, leadtimes, **kwargs)
    return np.sqrt(u ** 2 + v ** 2)

def wind_speed_ensavg(
    data: Dict[str, xr.Dataset], stations, reftimes, leadtimes, **kwargs
) -> xr.DataArray:
    ensemble = wind_speed(data, stations, reftimes, leadtimes, **kwargs)
    return ensemble.mean("realization")

def wind_speed_ensctrl(
    data: Dict[str, xr.Dataset], stations, reftimes, leadtimes, **kwargs
) -> xr.DataArray:
    ensemble = wind_speed(data, stations, reftimes, leadtimes, **kwargs)
    return ensemble.isel(realization=0, drop=True)


def wind_from_direction_ensctrl(
    data: Dict[str, xr.Dataset], stations, reftimes, leadtimes, **kwargs
) -> xr.DataArray:
    u = eastward_wind(data, stations, reftimes, leadtimes, **kwargs)
    v = northward_wind(data, stations, reftimes, leadtimes, **kwargs)
    out = (270 - 180 / np.pi * np.arctan2(v, u)) % 360
    out.attrs["units"] = "degrees"
    return out
```

this way nothing will be computed twice. The `callonce` decorator can store intermediate results with `pickle` in a temporary folder and read them when needed.

```python 
def callonce(func):
    @wraps(func)
    def wrapper(*args, **kwargs):
        cachedfile = pathlib.Path(kwargs["tmp_dir"]) / f"{func.__name__}"
        if not cachedfile.is_file():
            out = func(*args, **kwargs).load()
            with open(cachedfile, "wb") as f:
                pickle.dump(out, f, protocol=-1)
        else:
            with open(cachedfile, "rb") as f:
                out = pickle.load(f)
        return out
    return wrapper
```

### Use the `preproc` accessor only for indexing, selection, interpolation, alignment and complex statistics
If a computation is used only once, there's no reason it should be part of the accessor. I'd rather make it an explicit part of the computation. This is more transparent and also plays well with the caching system.
On the other hand, operations that are needed for many pipelines will remain in the `preproc` accessor.

**Comment by frazane** [comment]
1. yes, in the branch on which I am working I already defined a new decorator that does everything we need (f030bdb1d92b2c91965d082b9e10139fbbf05462). I would cache on disk to reduce RAM usage, and it's still quite fast with pickle. How could we handle the different arguments? My idea here for the moment is to use `tempfile` to create a temporary caching directory for each job/source.

2. Exactly. Methods that are intended to be used for specific features. I think `norm` could also be used explicitely I don't see any other use case on top of wind speed.

**Comment by frazane** [comment]
Solved by #16
